libpod/image/image.go

```diff
@@ -149,43 +210,344 @@ func (i *Image) ID() string {
+	}
+	_, _, err = imageSource.GetManifest(nil)
+	if err == nil {
+		return nil
+	}
+	return err
+}
+*/
+
+// GetImage retrieves an image matching the given name or hash from system
+// storage
+// If no matching image can be found, an error is returned
+func (ir *ImageRuntime) GetImage(image string) (*Image, error) {
+	var img *storage.Image
+	ref, err := is.Transport.ParseStoreReference(ir.store, image)
+	if err == nil {
+		img, err = is.Transport.GetStoreImage(ir.store, ref)
+	}
+	if err != nil {
+		img2, err2 := ir.store.Image(image)
+		if err2 != nil {
+			if ref == nil {
+				return nil, errors.Wrapf(err, "error parsing reference to image %q", image)
+			}
+			return nil, errors.Wrapf(err, "unable to locate image %q", image)
+		}
+		img = img2
+	}
+	newImage := ir.newFromStorage(img)
+	return newImage, nil
+}
+
+// GetImages retrieves all images present in storage
+// Filters can be provided which will determine which images are included in the
+// output. Multiple filters are handled by ANDing their output, so only images
+// matching all filters are included
+func (ir *ImageRuntime) GetImages(params *libpod.ImageFilterParams, filters ...libpod.ImageFilter) ([]*Image, error) {
+	var imagesFiltered []*storage.Image
+	var newImages []*Image
+
+	images, err := ir.store.Images()
 	if err != nil {
 		return nil, err
 	}
 
-	if decomposedImage.hasRegistry {
-		pullNames = append(pullNames, i.InputName)
-	} else {
-		registries, err := libpod.GetRegistries()
+	for _, img := range images {
+
+		info, err := ir.newFromStorage(&img).GetImageInspectInfo()
 		if err != nil {
 			return nil, err
 		}
-		for _, registry := range registries {
-			decomposedImage.registry = registry
-			pullNames = append(pullNames, decomposedImage.assemble())
+		var names []string
+		if len(img.Names) > 0 {
+			names = img.Names
+		} else {
+			names = append(names, "<none>")
+		}
+		for _, name := range names {
+			include := true
+			if params != nil {
+				params.ImageName = name
+			}
+			for _, filter := range filters {
+				include = include && filter(&img, info)
+			}
+
+			if include {
+				newImage := img
+				newImage.Names = []string{name}
+				imagesFiltered = append(imagesFiltered, &newImage)
+			}
 		}
 	}
-	return pullNames, nil
+
+	for _, i := range imagesFiltered {
+		newImages = append(newImages, ir.newFromStorage(i))
+	}
+	return newImages, nil
 }
 
-// pull is a temporary function for stage1 to be able to pull images during the image
-// resolution tests.  it will be replaced in stage2 with a more robust function.
-func (i *Image) pull(name string, writer io.Writer, r *libpod.Runtime) (string, error) {
-	options := libpod.CopyOptions{
-		Writer:              writer,
-		SignaturePolicyPath: r.GetConfig().SignaturePolicyPath,
+// getImageDigest creates an image object and uses the hex value of the digest as the image ID
+// for parsing the store reference
+func getImageDigest(src types.ImageReference, ctx *types.SystemContext) (string, error) {
```

Comment:
Can this be changed to work on a *Image instead? Or is it more of a general-purpose helper?

Reply:
more general purpose. going to leave for now